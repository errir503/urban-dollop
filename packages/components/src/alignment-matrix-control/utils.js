/**
 * External dependencies
 */
import { flattenDeep } from 'lodash';

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';

export const GRID = [
	[ 'top left', 'top center', 'top right' ],
	[ 'center left', 'center center', 'center right' ],
	[ 'bottom left', 'bottom center', 'bottom right' ],
];

// Stored as map as i18n __() only accepts strings (not variables)
export const ALIGNMENT_LABEL = {
	'top left': __( 'Top Left' ),
	'top center': __( 'Top Center' ),
	'top right': __( 'Top Right' ),
	'center left': __( 'Center Left' ),
	'center center': __( 'Center Center' ),
	'center right': __( 'Center Right' ),
	'bottom left': __( 'Bottom Left' ),
	'bottom center': __( 'Bottom Center' ),
	'bottom right': __( 'Bottom Right' ),
};

// Transforms GRID into a flat Array of values
export const ALIGNMENTS = flattenDeep( GRID );

/**
 * Parses and transforms an incoming value to better match the alignment values
 *
 * @param {string} value An alignment value to parse.
 *
 * @return {string} The parsed value.
 */
export function transformValue( value ) {
	const nextValue = value === 'center' ? 'center center' : value;

	return nextValue.replace( '-', ' ' );
}

/**
 * Creates an item ID based on a prefix ID and an alignment value.
 *
 * @param {string} prefixId An ID to prefix.
 * @param {string} value    An alignment value.
 *
 * @return {string} The item id.
 */
export function getItemId( prefixId, value ) {
	const valueId = transformValue( value ).replace( ' ', '-' );

	return `${ prefixId }-${ valueId }`;
}

/**
 * Retrieves the alignment index from a value.
 *
 * @param {string} alignment Value to check.
 *
 * @return {number} The index of a matching alignment.
 */
export function getAlignmentIndex( alignment = 'center' ) {
	const item = transformValue( alignment ).replace( '-', ' ' );
	const index = ALIGNMENTS.indexOf( item );

	return index > -1 ? index : undefined;
}
