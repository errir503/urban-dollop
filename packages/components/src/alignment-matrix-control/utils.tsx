/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';
/**
 * Internal dependencies
 */
import type { AlignmentMatrixControlValue } from './types';

export const GRID: AlignmentMatrixControlValue[][] = [
	[ 'top left', 'top center', 'top right' ],
	[ 'center left', 'center center', 'center right' ],
	[ 'bottom left', 'bottom center', 'bottom right' ],
];

// Stored as map as i18n __() only accepts strings (not variables)
export const ALIGNMENT_LABEL: Record< AlignmentMatrixControlValue, string > = {
	'top left': __( 'Top Left' ),
	'top center': __( 'Top Center' ),
	'top right': __( 'Top Right' ),
	'center left': __( 'Center Left' ),
	'center center': __( 'Center' ),
	center: __( 'Center' ),
	'center right': __( 'Center Right' ),
	'bottom left': __( 'Bottom Left' ),
	'bottom center': __( 'Bottom Center' ),
	'bottom right': __( 'Bottom Right' ),
};

// Transforms GRID into a flat Array of values.
export const ALIGNMENTS = GRID.flat();

/**
 * Parses and transforms an incoming value to better match the alignment values
 *
 * @param value An alignment value to parse.
 *
 * @return The parsed value.
 */
export function transformValue( value: AlignmentMatrixControlValue ) {
	const nextValue = value === 'center' ? 'center center' : value;

	return nextValue.replace( '-', ' ' ) as AlignmentMatrixControlValue;
}

/**
 * Creates an item ID based on a prefix ID and an alignment value.
 *
 * @param prefixId An ID to prefix.
 * @param value    An alignment value.
 *
 * @return The item id.
 */
export function getItemId(
	prefixId: string,
	value: AlignmentMatrixControlValue
) {
	const valueId = transformValue( value ).replace( ' ', '-' );

	return `${ prefixId }-${ valueId }`;
}

/**
 * Retrieves the alignment index from a value.
 *
 * @param alignment Value to check.
 *
 * @return The index of a matching alignment.
 */
export function getAlignmentIndex(
	alignment: AlignmentMatrixControlValue = 'center'
) {
	const item = transformValue( alignment );
	const index = ALIGNMENTS.indexOf( item );

	return index > -1 ? index : undefined;
}
