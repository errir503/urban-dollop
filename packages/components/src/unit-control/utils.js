/**
 * External dependencies
 */
import { isEmpty } from 'lodash';

/**
 * WordPress dependencies
 */
import { __, _x } from '@wordpress/i18n';
import { Platform } from '@wordpress/element';

const isWeb = Platform.OS === 'web';

const allUnits = {
	px: {
		value: 'px',
		label: isWeb ? 'px' : __( 'Pixels (px)' ),
		default: '',
		a11yLabel: __( 'Pixels (px)' ),
		step: 1,
	},
	percent: {
		value: '%',
		label: isWeb ? '%' : __( 'Percentage (%)' ),
		default: '',
		a11yLabel: __( 'Percent (%)' ),
		step: 0.1,
	},
	em: {
		value: 'em',
		label: isWeb ? 'em' : __( 'Relative to parent font size (em)' ),
		default: '',
		a11yLabel: _x( 'ems', 'Relative to parent font size (em)' ),
		step: 0.01,
	},
	rem: {
		value: 'rem',
		label: isWeb ? 'rem' : __( 'Relative to root font size (rem)' ),
		default: '',
		a11yLabel: _x( 'rems', 'Relative to root font size (rem)' ),
		step: 0.01,
	},
	vw: {
		value: 'vw',
		label: isWeb ? 'vw' : __( 'Viewport width (vw)' ),
		default: '',
		a11yLabel: __( 'Viewport width (vw)' ),
		step: 0.1,
	},
	vh: {
		value: 'vh',
		label: isWeb ? 'vh' : __( 'Viewport height (vh)' ),
		default: '',
		a11yLabel: __( 'Viewport height (vh)' ),
		step: 0.1,
	},
	vmin: {
		value: 'vmin',
		label: isWeb ? 'vmin' : __( 'Viewport smallest dimension (vmin)' ),
		default: '',
		a11yLabel: __( 'Viewport smallest dimension (vmin)' ),
		step: 0.1,
	},
	vmax: {
		value: 'vmax',
		label: isWeb ? 'vmax' : __( 'Viewport largest dimension (vmax)' ),
		default: '',
		a11yLabel: __( 'Viewport largest dimension (vmax)' ),
		step: 0.1,
	},
	ch: {
		value: 'ch',
		label: isWeb ? 'ch' : __( 'Width of the zero (0) character (ch)' ),
		default: '',
		a11yLabel: __( 'Width of the zero (0) character (ch)' ),
		step: 0.01,
	},
	ex: {
		value: 'ex',
		label: isWeb ? 'ex' : __( 'x-height of the font (ex)' ),
		default: '',
		a11yLabel: __( 'x-height of the font (ex)' ),
		step: 0.01,
	},
	cm: {
		value: 'cm',
		label: isWeb ? 'cm' : __( 'Centimeters (cm)' ),
		default: '',
		a11yLabel: __( 'Centimeters (cm)' ),
		step: 0.001,
	},
	mm: {
		value: 'mm',
		label: isWeb ? 'mm' : __( 'Millimeters (mm)' ),
		default: '',
		a11yLabel: __( 'Millimeters (mm)' ),
		step: 0.1,
	},
	in: {
		value: 'in',
		label: isWeb ? 'in' : __( 'Inches (in)' ),
		default: '',
		a11yLabel: __( 'Inches (in)' ),
		step: 0.001,
	},
	pc: {
		value: 'pc',
		label: isWeb ? 'pc' : __( 'Picas (pc)' ),
		default: '',
		a11yLabel: __( 'Picas (pc)' ),
		step: 1,
	},
	pt: {
		value: 'pt',
		label: isWeb ? 'pt' : __( 'Points (pt)' ),
		default: '',
		a11yLabel: __( 'Points (pt)' ),
		step: 1,
	},
};

/**
 * An array of all available CSS length units.
 */
export const ALL_CSS_UNITS = Object.values( allUnits );

/**
 * Units of measurements. `a11yLabel` is used by screenreaders.
 */
export const CSS_UNITS = [
	allUnits.px,
	allUnits.percent,
	allUnits.em,
	allUnits.rem,
	allUnits.vw,
	allUnits.vh,
];

export const DEFAULT_UNIT = allUnits.px;

/**
 * Handles legacy value + unit handling.
 * This component use to manage both incoming value and units separately.
 *
 * Moving forward, ideally the value should be a string that contains both
 * the value and unit, example: '10px'
 *
 * @param {number|string} value Value
 * @param {string}        unit  Unit value
 * @param {Array<Object>} units Units to derive from.
 * @return {Array<number, string>} The extracted number and unit.
 */
export function getParsedValue( value, unit, units ) {
	const initialValue = unit ? `${ value }${ unit }` : value;

	return parseUnit( initialValue, units );
}

/**
 * Checks if units are defined.
 *
 * @param {any} units Units to check.
 * @return {boolean} Whether units are defined.
 */
export function hasUnits( units ) {
	return ! isEmpty( units ) && units !== false;
}

/**
 * Parses a number and unit from a value.
 *
 * @param {string}        initialValue Value to parse
 * @param {Array<Object>} units        Units to derive from.
 * @return {Array<number, string>} The extracted number and unit.
 */
export function parseUnit( initialValue, units = ALL_CSS_UNITS ) {
	const value = String( initialValue ).trim();

	let num = parseFloat( value, 10 );
	num = isNaN( num ) ? '' : num;

	const unitMatch = value.match( /[\d.\-\+]*\s*(.*)/ )[ 1 ];

	let unit = unitMatch !== undefined ? unitMatch : '';
	unit = unit.toLowerCase();

	if ( hasUnits( units ) ) {
		const match = units.find( ( item ) => item.value === unit );
		unit = match?.value;
	} else {
		unit = DEFAULT_UNIT.value;
	}

	return [ num, unit ];
}

/**
 * Parses a number and unit from a value. Validates parsed value, using fallback
 * value if invalid.
 *
 * @param {number|string} next          The next value.
 * @param {Array<Object>} units         Units to derive from.
 * @param {number|string} fallbackValue The fallback value.
 * @param {string}        fallbackUnit  The fallback value.
 * @return {Array<number, string>} The extracted number and unit.
 */
export function getValidParsedUnit( next, units, fallbackValue, fallbackUnit ) {
	const [ parsedValue, parsedUnit ] = parseUnit( next, units );
	let baseValue = parsedValue;
	let baseUnit;

	if ( isNaN( parsedValue ) || parsedValue === '' ) {
		baseValue = fallbackValue;
	}

	baseUnit = parsedUnit || fallbackUnit;

	/**
	 * If no unit is found, attempt to use the first value from the collection
	 * of units as a default fallback.
	 */
	if ( hasUnits( units ) && ! baseUnit ) {
		baseUnit = units[ 0 ]?.value;
	}

	return [ baseValue, baseUnit ];
}

/**
 * Takes a unit value and finds the matching accessibility label for the
 * unit abbreviation.
 *
 * @param {string} unit Unit value (example: px)
 * @return {string} a11y label for the unit abbreviation
 */
export function parseA11yLabelForUnit( unit ) {
	const match = ALL_CSS_UNITS.find( ( item ) => item.value === unit );
	return match?.a11yLabel ? match?.a11yLabel : match?.value;
}

/**
 * Filters available units based on values defined by settings.
 *
 * @param {Array} settings Collection of preferred units.
 * @param {Array} units    Collection of available units.
 *
 * @return {Array} Filtered units based on settings.
 */
export function filterUnitsWithSettings( settings = [], units = [] ) {
	return units.filter( ( unit ) => {
		return settings.includes( unit.value );
	} );
}

/**
 * Custom hook to retrieve and consolidate units setting from add_theme_support().
 * TODO: ideally this hook shouldn't be needed
 * https://github.com/WordPress/gutenberg/pull/31822#discussion_r633280823
 *
 * @param {Object} args                                  An object containing units, settingPath & defaultUnits.
 * @param {Array<Object>|undefined} args.units           Collection of available units.
 * @param {Array<string>|undefined} args.availableUnits  The setting path. Defaults to 'spacing.units'.
 * @param {Object|undefined}        args.defaultValues   Collection of default values for defined units. Example: { px: '350', em: '15' }.
 *
 * @return {Array|boolean} Filtered units based on settings.
 */
export const useCustomUnits = ( { units, availableUnits, defaultValues } ) => {
	units = units || ALL_CSS_UNITS;
	const usedUnits = filterUnitsWithSettings(
		! availableUnits ? [] : availableUnits,
		units
	);

	if ( defaultValues ) {
		usedUnits.forEach( ( unit, i ) => {
			if ( defaultValues[ unit.value ] ) {
				usedUnits[ i ].default = defaultValues[ unit.value ];
			}
		} );
	}

	return usedUnits.length === 0 ? false : usedUnits;
};
