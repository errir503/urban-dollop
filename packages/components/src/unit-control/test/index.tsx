/**
 * External dependencies
 */
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

/**
 * WordPress dependencies
 */
import { useState } from '@wordpress/element';

/**
 * Internal dependencies
 */
import UnitControl from '..';
import { parseQuantityAndUnitFromRawValue } from '../utils';
import type { UnitControlOnChangeCallback } from '../types';

const getInput = ( {
	isInputTypeText = false,
}: {
	isInputTypeText?: boolean;
} = {} ) =>
	screen.getByRole(
		isInputTypeText ? 'textbox' : 'spinbutton'
	) as HTMLInputElement;
const getSelect = () => screen.getByRole( 'combobox' ) as HTMLSelectElement;
const getSelectOptions = () =>
	screen.getAllByRole( 'option' ) as HTMLOptionElement[];

const ControlledSyncUnits = () => {
	const [ state, setState ] = useState( {
		valueA: '',
		valueB: '',
	} );

	// Keep the unit sync'd between the two `UnitControl` instances.
	const onUnitControlChange = (
		fieldName: 'valueA' | 'valueB',
		newValue?: string | number
	) => {
		const parsedQuantityAndUnit =
			parseQuantityAndUnitFromRawValue( newValue );
		const quantity = parsedQuantityAndUnit[ 0 ];

		if ( ! Number.isFinite( quantity ) ) {
			return;
		}

		const newUnit = parsedQuantityAndUnit[ 1 ];

		const nextState = {
			...state,
			[ fieldName ]: newValue,
		};

		Object.entries( state ).forEach( ( [ stateProp, stateValue ] ) => {
			const [ stateQuantity, stateUnit ] =
				parseQuantityAndUnitFromRawValue( stateValue );

			if ( stateProp !== fieldName && stateUnit !== newUnit ) {
				nextState[
					stateProp as 'valueA' | 'valueB'
				] = `${ stateQuantity }${ newUnit }`;
			}
		} );

		setState( nextState );
	};

	return (
		<>
			<UnitControl
				label="Field A"
				value={ state.valueA }
				onChange={ ( v ) => onUnitControlChange( 'valueA', v ) }
			/>
			<UnitControl
				label="Field B"
				value={ state.valueB }
				onChange={ ( v ) => onUnitControlChange( 'valueB', v ) }
			/>
		</>
	);
};

describe( 'UnitControl', () => {
	beforeEach( () => {
		jest.useFakeTimers();
	} );

	afterEach( () => {
		jest.runOnlyPendingTimers();
		jest.useRealTimers();
	} );

	describe( 'Basic rendering', () => {
		it( 'should render', () => {
			render( <UnitControl /> );
			const input = getInput();
			const select = getSelect();

			expect( input ).toBeInTheDocument();
			expect( select ).toBeInTheDocument();
		} );

		it( 'should render custom className', () => {
			const { container: withoutClassName } = render( <UnitControl /> );

			const { container: withClassName } = render(
				<UnitControl className="hello" />
			);

			expect(
				withoutClassName.querySelector( '.components-unit-control' )
			).not.toHaveClass( 'hello' );
			expect(
				withClassName.querySelector( '.components-unit-control' )
			).toHaveClass( 'hello' );
		} );

		it( 'should not render select, if units are disabled', () => {
			render( <UnitControl value="3em" units={ [] } /> );
			const input = getInput();
			// Using `queryByRole` instead of `getSelect` because we need to test
			// for this element NOT to be in the document.
			const select = screen.queryByRole( 'combobox' );

			expect( input ).toBeInTheDocument();
			expect( select ).not.toBeInTheDocument();
		} );

		it( 'should render label if single units', () => {
			render( <UnitControl units={ [ { value: '%', label: '%' } ] } /> );

			const select = screen.queryByRole( 'combobox' );
			const label = screen.getByText( '%' );

			expect( select ).not.toBeInTheDocument();
			expect( label ).toBeInTheDocument();
		} );
	} );

	describe( 'Value', () => {
		it( 'should update value on change', async () => {
			const user = userEvent.setup( {
				advanceTimers: jest.advanceTimersByTime,
			} );

			let state = '50px';
			const setState = jest.fn( ( value ) => ( state = value ) );

			render( <UnitControl value={ state } onChange={ setState } /> );

			const input = getInput();
			await user.clear( input );
			await user.type( input, '62' );

			// 3 times:
			// - 1: clear
			// - 2: type '6'
			// - 3: type '62'
			expect( setState ).toHaveBeenCalledTimes( 3 );
			expect( state ).toBe( '62px' );
		} );

		it( 'should increment value on UP press', async () => {
			const user = userEvent.setup( {
				advanceTimers: jest.advanceTimersByTime,
			} );

			let state: string | undefined = '50px';
			const setState: UnitControlOnChangeCallback = ( nextState ) =>
				( state = nextState );

			render( <UnitControl value={ state } onChange={ setState } /> );

			const input = getInput();
			await user.type( input, '{ArrowUp}' );

			expect( state ).toBe( '51px' );
		} );

		it( 'should increment value on UP + SHIFT press, with step', async () => {
			const user = userEvent.setup( {
				advanceTimers: jest.advanceTimersByTime,
			} );

			let state: string | undefined = '50px';
			const setState: UnitControlOnChangeCallback = ( nextState ) =>
				( state = nextState );

			render( <UnitControl value={ state } onChange={ setState } /> );

			const input = getInput();
			await user.type( input, '{Shift>}{ArrowUp}{/Shift}' );

			expect( state ).toBe( '60px' );
		} );

		it( 'should decrement value on DOWN press', async () => {
			const user = userEvent.setup( {
				advanceTimers: jest.advanceTimersByTime,
			} );

			let state: string | number | undefined = 50;
			const setState: UnitControlOnChangeCallback = ( nextState ) =>
				( state = nextState );

			render( <UnitControl value={ state } onChange={ setState } /> );

			const input = getInput();
			await user.type( input, '{ArrowDown}' );

			expect( state ).toBe( '49px' );
		} );

		it( 'should decrement value on DOWN + SHIFT press, with step', async () => {
			const user = userEvent.setup( {
				advanceTimers: jest.advanceTimersByTime,
			} );

			let state: string | number | undefined = 50;
			const setState: UnitControlOnChangeCallback = ( nextState ) =>
				( state = nextState );

			render( <UnitControl value={ state } onChange={ setState } /> );

			const input = getInput();
			await user.type( input, '{Shift>}{ArrowDown}{/Shift}' );

			expect( state ).toBe( '40px' );
		} );

		it( 'should cancel change when ESCAPE key is pressed', async () => {
			const user = userEvent.setup( {
				advanceTimers: jest.advanceTimersByTime,
			} );

			let state: string | number | undefined = 50;
			const setState: UnitControlOnChangeCallback = ( nextState ) =>
				( state = nextState );

			render(
				<UnitControl
					value={ state }
					onChange={ setState }
					isPressEnterToChange
				/>
			);

			// Input type is `text` when the `isPressEnterToChange` prop is passed
			const input = getInput( { isInputTypeText: true } );
			await user.clear( input );
			await user.type( input, '300px' );

			expect( input.value ).toBe( '300px' );
			expect( state ).toBe( 50 );

			await user.keyboard( '{Escape}' );

			expect( input.value ).toBe( '50' );
			expect( state ).toBe( 50 );
		} );

		it( 'should run onBlur callback when quantity input is blurred', async () => {
			const user = userEvent.setup( {
				advanceTimers: jest.advanceTimersByTime,
			} );

			const onChangeSpy = jest.fn();
			const onBlurSpy = jest.fn();

			let state: string | undefined = '33%';
			const setState: UnitControlOnChangeCallback = ( nextState ) => {
				onChangeSpy( nextState );
				state = nextState;
			};

			render(
				<UnitControl
					value={ state }
					onChange={ setState }
					onBlur={ onBlurSpy }
				/>
			);

			const input = getInput();
			await user.clear( input );
			await user.type( input, '41' );

			expect( onChangeSpy ).toHaveBeenCalledTimes( 3 );
			expect( onChangeSpy ).toHaveBeenLastCalledWith( '41%' );

			// Clicking document.body to trigger a blur event on the input.
			await user.click( document.body );

			expect( onBlurSpy ).toHaveBeenCalledTimes( 1 );
		} );

		it( 'should invoke onChange and onUnitChange callbacks when isPressEnterToChange is true and the component is blurred with an uncommitted value', async () => {
			const user = userEvent.setup( {
				advanceTimers: jest.advanceTimersByTime,
			} );

			const onUnitChangeSpy = jest.fn();
			const onChangeSpy = jest.fn();

			let state: string | undefined = '15px';
			const setState: UnitControlOnChangeCallback = ( nextState ) => {
				onChangeSpy( nextState );
				state = nextState;
			};

			render(
				<UnitControl
					value={ state }
					onChange={ setState }
					onUnitChange={ onUnitChangeSpy }
					isPressEnterToChange
				/>
			);

			// Input type is `text` when the `isPressEnterToChange` prop is passed
			const input = getInput( { isInputTypeText: true } );
			await user.clear( input );
			await user.type( input, '41vh' );

			// This is because `isPressEnterToChange` is `true`
			expect( onChangeSpy ).not.toHaveBeenCalled();
			expect( onUnitChangeSpy ).not.toHaveBeenCalled();

			// Clicking document.body to trigger a blur event on the input.
			await user.click( document.body );

			expect( onChangeSpy ).toHaveBeenCalledTimes( 1 );
			expect( onChangeSpy ).toHaveBeenLastCalledWith( '41vh' );

			expect( onUnitChangeSpy ).toHaveBeenCalledTimes( 1 );
			expect( onUnitChangeSpy ).toHaveBeenLastCalledWith(
				'vh',
				expect.anything()
			);
		} );

		it( 'should update value correctly when typed and blurred when a single unit is passed', async () => {
			const user = userEvent.setup( {
				advanceTimers: jest.advanceTimersByTime,
			} );

			const onChangeSpy = jest.fn();
			render(
				<>
					<button>Click me</button>
					<UnitControl
						units={ [ { value: '%', label: '%' } ] }
						onChange={ onChangeSpy }
					/>
				</>
			);

			const input = getInput();
			await user.type( input, '62' );

			expect( onChangeSpy ).toHaveBeenLastCalledWith(
				'62%',
				expect.anything()
			);

			// Start counting again calls to `onChangeSpy`.
			onChangeSpy.mockClear();

			// Clicking on the button should cause the `onBlur` callback to fire.
			const button = screen.getByRole( 'button' );
			await user.click( button );

			expect( onChangeSpy ).not.toHaveBeenCalled();
		} );
	} );

	describe( 'Unit', () => {
		it( 'should update unit value on change', async () => {
			const user = userEvent.setup( {
				advanceTimers: jest.advanceTimersByTime,
			} );

			let state: string | undefined = '14rem';
			const setState: UnitControlOnChangeCallback = ( nextState ) =>
				( state = nextState );

			const spy = jest.fn();

			render(
				<UnitControl
					value={ state }
					onChange={ setState }
					onUnitChange={ spy }
				/>
			);

			const select = getSelect();
			await user.selectOptions( select, [ 'px' ] );

			expect( spy ).toHaveBeenCalledWith( 'px', expect.anything() );
			expect( state ).toBe( '14px' );
		} );

		it( 'should render customized units, if defined', () => {
			const units = [
				{ value: 'pt', label: 'pt', default: 0 },
				{ value: 'vmax', label: 'vmax', default: 10 },
			];

			render( <UnitControl units={ units } /> );

			const options = getSelectOptions();

			expect( options.length ).toBe( 2 );

			const [ pt, vmax ] = options;

			expect( pt.value ).toBe( 'pt' );
			expect( vmax.value ).toBe( 'vmax' );
		} );

		it( 'should reset value on unit change, if unit has default value', async () => {
			const user = userEvent.setup( {
				advanceTimers: jest.advanceTimersByTime,
			} );

			let state: string | number | undefined = 50;
			const setState: UnitControlOnChangeCallback = ( nextState ) =>
				( state = nextState );

			const units = [
				{ value: 'pt', label: 'pt', default: 25 },
				{ value: 'vmax', label: 'vmax', default: 75 },
			];

			render(
				<UnitControl
					isResetValueOnUnitChange
					units={ units }
					onChange={ setState }
					value={ state }
				/>
			);

			const select = getSelect();
			await user.selectOptions( select, [ 'vmax' ] );

			expect( state ).toBe( '75vmax' );

			await user.selectOptions( select, [ 'pt' ] );

			expect( state ).toBe( '25pt' );
		} );

		it( 'should not reset value on unit change, if disabled', async () => {
			const user = userEvent.setup( {
				advanceTimers: jest.advanceTimersByTime,
			} );

			let state: string | number | undefined = 50;
			const setState: UnitControlOnChangeCallback = ( nextState ) =>
				( state = nextState );

			const units = [
				{ value: 'pt', label: 'pt', default: 25 },
				{ value: 'vmax', label: 'vmax', default: 75 },
			];

			render(
				<UnitControl
					isResetValueOnUnitChange={ false }
					value={ state }
					units={ units }
					onChange={ setState }
				/>
			);

			const select = getSelect();
			await user.selectOptions( select, [ 'vmax' ] );

			expect( state ).toBe( '50vmax' );

			await user.selectOptions( select, [ 'pt' ] );

			expect( state ).toBe( '50pt' );
		} );

		it( 'should set correct unit if single units', async () => {
			const user = userEvent.setup( {
				advanceTimers: jest.advanceTimersByTime,
			} );

			let state: string | undefined = '50%';
			const setState: UnitControlOnChangeCallback = ( value ) =>
				( state = value );

			render(
				<UnitControl
					value={ state }
					units={ [ { value: '%', label: '%' } ] }
					onChange={ setState }
				/>
			);

			const input = getInput();
			await user.clear( input );
			await user.type( input, '62' );

			expect( state ).toBe( '62%' );
		} );

		it( 'should update unit value when a new raw value is passed', async () => {
			const user = userEvent.setup( {
				advanceTimers: jest.advanceTimersByTime,
			} );

			render( <ControlledSyncUnits /> );

			const [ inputA, inputB ] = screen.getAllByRole( 'spinbutton' );
			const [ selectA, selectB ] = screen.getAllByRole( 'combobox' );

			const [ remOptionA ] = screen.getAllByRole( 'option', {
				name: 'rem',
			} );
			const [ , vwOptionB ] = screen.getAllByRole( 'option', {
				name: 'vw',
			} );

			await user.type( inputA, '55' );

			await user.type( inputB, '14' );

			await user.selectOptions( selectA, remOptionA );

			expect( selectB ).toHaveValue( 'rem' );
			expect( selectA ).toHaveValue( 'rem' );

			await user.selectOptions( selectB, vwOptionB );

			expect( selectA ).toHaveValue( 'vw' );
			expect( selectB ).toHaveValue( 'vw' );
		} );

		it( 'should maintain the chosen non-default unit when value is cleared', async () => {
			const user = userEvent.setup( {
				advanceTimers: jest.advanceTimersByTime,
			} );

			const units = [
				{ value: 'pt', label: 'pt' },
				{ value: 'vmax', label: 'vmax' },
			];

			render( <UnitControl units={ units } value="5" /> );

			const select = getSelect();
			await user.selectOptions( select, [ 'vmax' ] );

			const input = getInput();
			await user.clear( input );

			expect( select ).toHaveValue( 'vmax' );
		} );

		it( 'should run onBlur callback when the unit select is blurred', async () => {
			const user = userEvent.setup( {
				advanceTimers: jest.advanceTimersByTime,
			} );

			const onUnitChangeSpy = jest.fn();
			const onBlurSpy = jest.fn();

			render(
				<UnitControl
					value="15px"
					onUnitChange={ onUnitChangeSpy }
					onBlur={ onBlurSpy }
				/>
			);

			const select = getSelect();
			await user.selectOptions( select, [ 'em' ] );

			expect( onUnitChangeSpy ).toHaveBeenCalledTimes( 1 );
			expect( onUnitChangeSpy ).toHaveBeenLastCalledWith(
				'em',
				expect.anything()
			);

			// Clicking document.body to trigger a blur event on the input.
			await user.click( document.body );

			expect( onBlurSpy ).toHaveBeenCalledTimes( 1 );
		} );
	} );

	describe( 'Unit Parser', () => {
		it( 'should parse unit from input', async () => {
			const user = userEvent.setup( {
				advanceTimers: jest.advanceTimersByTime,
			} );

			let state = '10px';
			const setState = jest.fn( ( nextState ) => ( state = nextState ) );

			render(
				<UnitControl
					value={ state }
					onChange={ setState }
					isPressEnterToChange
				/>
			);

			// Input type is `text` when the `isPressEnterToChange` prop is passed
			const input = getInput( { isInputTypeText: true } );
			await user.clear( input );
			await user.type( input, '55 em' );
			await user.keyboard( '{Enter}' );

			expect( state ).toBe( '55em' );
		} );

		it( 'should parse PX unit from input', async () => {
			const user = userEvent.setup( {
				advanceTimers: jest.advanceTimersByTime,
			} );

			let state = '10px';
			const setState = jest.fn( ( nextState ) => ( state = nextState ) );

			render(
				<UnitControl
					value={ state }
					onChange={ setState }
					isPressEnterToChange
				/>
			);

			// Input type is `text` when the `isPressEnterToChange` prop is passed
			const input = getInput( { isInputTypeText: true } );
			await user.clear( input );
			await user.type( input, '61   PX' );
			await user.keyboard( '{Enter}' );

			expect( state ).toBe( '61px' );
		} );

		it( 'should parse EM unit from input', async () => {
			const user = userEvent.setup( {
				advanceTimers: jest.advanceTimersByTime,
			} );

			let state = '10px';
			const setState = jest.fn( ( nextState ) => ( state = nextState ) );

			render(
				<UnitControl
					value={ state }
					onChange={ setState }
					isPressEnterToChange
				/>
			);

			// Input type is `text` when the `isPressEnterToChange` prop is passed
			const input = getInput( { isInputTypeText: true } );
			await user.clear( input );
			await user.type( input, '55 em' );
			await user.keyboard( '{Enter}' );

			expect( state ).toBe( '55em' );
		} );

		it( 'should parse % unit from input', async () => {
			const user = userEvent.setup( {
				advanceTimers: jest.advanceTimersByTime,
			} );

			let state = '10px';
			const setState = jest.fn( ( nextState ) => ( state = nextState ) );

			render(
				<UnitControl
					value={ state }
					onChange={ setState }
					isPressEnterToChange
				/>
			);

			// Input type is `text` when the `isPressEnterToChange` prop is passed
			const input = getInput( { isInputTypeText: true } );
			await user.clear( input );
			await user.type( input, '-10  %' );
			await user.keyboard( '{Enter}' );

			expect( state ).toBe( '-10%' );
		} );

		it( 'should parse REM unit from input', async () => {
			const user = userEvent.setup( {
				advanceTimers: jest.advanceTimersByTime,
			} );

			let state = '10px';
			const setState = jest.fn( ( nextState ) => ( state = nextState ) );

			render(
				<UnitControl
					value={ state }
					onChange={ setState }
					isPressEnterToChange
				/>
			);

			// Input type is `text` when the `isPressEnterToChange` prop is passed
			const input = getInput( { isInputTypeText: true } );
			await user.clear( input );
			await user.type( input, '123       rEm  ' );
			await user.keyboard( '{Enter}' );

			expect( state ).toBe( '123rem' );
		} );

		it( 'should update unit after initial render and with new unit prop', async () => {
			const { rerender } = render( <UnitControl value={ '10%' } /> );

			const select = getSelect();

			expect( select.value ).toBe( '%' );

			rerender( <UnitControl value={ '20vh' } /> );

			expect( select.value ).toBe( 'vh' );
		} );

		it( 'should fallback to default unit if parsed unit is invalid', () => {
			render( <UnitControl value={ '10null' } /> );

			expect( getSelect().value ).toBe( 'px' );
		} );

		it( 'should display valid CSS unit when not explicitly included in units list', () => {
			render(
				<UnitControl
					value={ '10%' }
					units={ [
						{ value: 'px', label: 'px' },
						{ value: 'em', label: 'em' },
					] }
				/>
			);

			const select = getSelect();
			const options = getSelectOptions();

			expect( select.value ).toBe( '%' );
			expect( options.length ).toBe( 3 );
		} );
	} );
} );
