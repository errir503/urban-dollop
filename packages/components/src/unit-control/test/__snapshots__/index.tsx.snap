// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnitControl Basic rendering should render custom className 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ -1,10 +1,10 @@
  <div
    class="components-unit-control-wrapper css-aa2xc3-Root e1bagdl33"
  >
    <div
-     class="components-flex components-input-control components-number-control components-unit-control e1bagdl32 ep48uk90 em5sgkm7 css-18wzek1-View-Flex-sx-Base-sx-Items-ItemsColumn-Root-rootLabelPositionStyles-Input-ValueInput-arrowStyles-paddingStyles em57xhy0"
+     class="components-flex components-input-control components-number-control components-unit-control hello e1bagdl32 ep48uk90 em5sgkm7 css-18wzek1-View-Flex-sx-Base-sx-Items-ItemsColumn-Root-rootLabelPositionStyles-Input-ValueInput-arrowStyles-paddingStyles em57xhy0"
      data-wp-c16t="true"
      data-wp-component="Flex"
    >
      <div
        class="components-flex-item em5sgkm3 css-1fmchc6-View-Item-sx-Base-LabelWrapper em57xhy0"
@@ -15,11 +15,11 @@
        class="components-input-control__container css-1sy20aj-Container-containerDisabledStyles-containerMarginStyles-containerWidthStyles em5sgkm6"
      >
        <input
          autocomplete="off"
          class="components-input-control__input css-gzm6eu-Input-dragStyles-fontSizeStyles-sizeStyles em5sgkm5"
-         id="inspector-input-control-1"
+         id="inspector-input-control-2"
          inputmode="numeric"
          max="Infinity"
          min="-Infinity"
          step="1"
          type="number"
`;
